lms/services/exceptions.py

```diff
@@ -91,6 +91,49 @@ def __str__(self):
         return repr(self)
 
 
+class ExternalAsyncRequestError(ExternalRequestError):
```

Comment:
I wonder whether the subclassing is worthwhile here (potentially coupling the two classes together), or should they just be two completely separate classes that happen to provide the same interface?

Reply:
Sub-classing gives us the `__repr__`  of `ExternalRequestError`.

We could try to cram everything on one class but I reckon it could end up being more confusing, it will have 2 or 3 conditionals in each method.

Reply:
I think we should avoid `ExternalAsyncRequestError` subclassing `ExternalRequestError` for a couple of reasons:

First, semantically it means that an `ExternalAsyncRequestError` **is a** `ExternalRequestError` and I'm not sure that we want that. For example `isinstance(async_external_request_error, ExternalRequestError)` will be `True`, `catch ExternalRequestError` will catch `AsyncExternalRequestError`'s, etc.

Second, it means that the `ExternalRequestError` class is now part of the `ExternalAsyncRequestError`. You can no longer change `ExternalRequestError` without changing `ExternalAsyncRequestError` as well. For example if I add a new method to `ExternalRequestError` (which a PR that I'm working on does) `ExternalAsyncRequestError` will inherit that method but the inherited method probably won't work if called on an `ExternalAsyncRequestError` instance because `request` and `response` will be different types of object than with `ExternalRequestError`.

I think there's a high chance of making a change to `ExternalRequestError` and not realising that `ExternalAsyncRequestError` is a subclass and you've affected or broken `ExternalAsyncRequestError`.

I think there may already be some unintended side-effects of the inheritance. For example an `ExternalAsyncRequestError` object has `request` and `validation_errors` attributes whose values are `None`.

I think we want to keep these two classes decoupled from each other but providing the same interface. The simplest way for them to share code might be for them to use shared helper functions?

```python
class ExternalRequestError(Exception):
    ...

    def __repr__(self):
        return _repr(self)

    def __str__(self):
        return repr(self)


class ExternalAsyncRequestError(Exception):
    ...

    def __repr__(self):
        return _repr(self)

    def __str__(self):
        return repr(self)


def _repr(exception):
    ...
```

Reply:
I think `ExternalAsyncRequestError` being a `ExternalRequestError` from OOP PoV was fine but I agree that that mostly ties them together implementation wise which could be a problem.

Separated them and added `ExternalAsyncRequestError` to the same exception view from now.